### ISSUE TITLE
Changed behavior of singledispatch in Python 3.12

### ISSUE THREAD

**Issue by AdrianSosic** [python/cpython#117097]
# Bug report

### Bug description:

It seems the behavior of `functools.singledispatch` changed in Python 3.12 in situations where a `typing.Protocol` is involved. 

When executing the following code on 3.12, `MyClass` no longer gets dispatched as via the default but as `MyProtocol`. The problem seems to be that `issubclass` now behaves differently when called on protocols. The issue was noticed [here](https://github.com/python-attrs/cattrs/issues/525) where it caused downstream problems.

```python
from typing import Protocol
from functools import singledispatch


class MyProtocol(Protocol):
    pass

class MyClass:
    pass

s = singledispatch(lambda x: "default")
s.register(MyProtocol, lambda x: "protocol")

print(s(MyClass))  # yields "default" on 3.11 but "protocol" on 3.12
```

### CPython versions tested on:

3.11, 3.12

### Operating systems tested on:

macOS

**Comment by sobolevn** [comment]
cc @AlexWaygood

**Comment by AlexWaygood** [comment]
Okay, I strongly suspect this is because of this delightful part of `typing.py`:

https://github.com/python/cpython/blob/8ad88984200b2ccddc0a08229dd2f4c14d1a71fc/Lib/typing.py#L1784-L1790

I'll investigate more tomorrow.

**Comment by AlexWaygood** [comment]
Thanks for the clear bug report, @AdrianSosic!

**Comment by AdrianSosic** [comment]
Hi @AlexWaygood, just wanted to ask if you could already find some time to figure out what exactly causes the problem? 🙃

**Comment by AlexWaygood** [comment]
Sorry, I haven't had a chance to look at this properly yet. But thanks for the ping. Please feel free to ping me again if I haven't responded in ~2 weeks :-)

**Comment by AlexWaygood** [comment]
Thanks, you're not annoying me!

**Comment by AlexWaygood** [comment]
If I haven't got to it by then, I'll definitely look at this at the PyCon Sprints (week commencing 20th May)

**Comment by AlexWaygood** [comment]
At least two commits changed the behaviour of this snippet between Python 3.11 and Python 3.12. The first commit that changed behaviour was b27fe67f3c643e174c3619b669228ef34b6d87ee. Prior to this commit, the snippet printed `"default"`; after this commit, the snippet raised `TypeError`. I'm now bisecting to see which commit after this changed behaviour so that it no longer raised `TypeError` and started printing `"protocol"`.

**Comment by AlexWaygood** [comment]
My feeling is that the new behaviour is probably correct. Calling `isinstance()` and `issubclass()` against a protocol that isn't decorated with `@runtime_checkable` isn't usually allowed, but we make an explicit exception for the `functools` module. But if you create an empty protocol -- as you're doing here -- and decorate it with `@runtime_checkable`, then you'll see that all objects are considered instances of that protocol. This has been the case since runtime-checkable protocols were originally introduced in Python 3.8:

```pycon
Python 3.8.18 (default, Feb 15 2024, 19:36:58) 
[Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import Protocol, runtime_checkable
>>> @runtime_checkable
... class EmptyProtocol(Protocol): pass
... 
>>> isinstance([], EmptyProtocol)
True
>>> isinstance(object(), EmptyProtocol)
True
>>> issubclass(dict, EmptyProtocol)
True
>>> issubclass(object, EmptyProtocol)
True
```

As such, it makes sense that `s.register(MyProtocol, lambda x: "protocol")` would lead to `lambda x: "protocol"` being selected as the relevant function variant for any arbitrary object being passed in. Non-empty protocols also appear to be behaving correctly:

```pycon
>>> from functools import singledispatch
>>> s = singledispatch(lambda x: "default")
>>> from typing import SupportsIndex
>>> _ = s.register(SupportsIndex, lambda x: "supports index")
>>> s('foo')
'default'
>>> s({})
'default'
>>> s(432)
'supports index'
```

So, I think the only thing to do here is to add a regression test, since it appears that there was a bug on Python <=3.11 that was accidentally fixed. @carljm / @JelleZijlstra, do you agree?

**Comment by JelleZijlstra** [comment]
I agree that everything should match an empty protocol and therefore the new behavior is correct.

However, it seems wrong that singledispatch allows dispatching on a non-runtime checkable protocol. I tried locally and `test_functools` and `test_typing` both pass if I remove functools from `_allow_reckless_class_checks`. I wonder if we can remove it.

**Comment by AlexWaygood** [comment]
> However, it seems wrong that singledispatch allows dispatching on a non-runtime checkable protocol. I tried locally and `test_functools` and `test_typing` both pass if I remove functools from `_allow_reckless_class_checks`. I wonder if we can remove it.

I agree that this does seem wrong. However, changing it will break users like @AdrianSosic who have been relying on the existing behaviour where it was allowed. I don't think we can change this without a deprecation period.
